nlp/optimize/tron.py

```diff
@@ -319,13 +315,17 @@ def projected_linesearch(self, x, l, u, g, d, H, alpha=1.0):
 
         where H=Hᵀ and g is a vector. Given the parameter α, the step is
 
-           s[α] = P[x + α d] - x,
+           s[α] = s₀ + P[x + α d] - x,
```

Comment:
You raise a good point here. My thinking was that I needed s0 in the function to check that the sum of s0 and the new direction was still a descent direction. However, the code is a lot more complicated than before (and probably less efficient) as a result.